dbms/src/Storages/MergeTree/MergeTreePartsMover.cpp

```diff
@@ -94,35 +94,91 @@ bool MergeTreePartsMover::selectPartsForMove(
     {
         for (const auto & disk : volumes[i]->disks)
         {
-            UInt64 required_available_space = disk->getTotalSpace() * policy->getMoveFactor();
+            UInt64 required_maximum_available_space = disk->getTotalSpace() * policy->getMoveFactor();
             UInt64 unreserved_space = disk->getUnreservedSpace();
 
-            if (required_available_space > unreserved_space)
-                need_to_move.emplace(disk, required_available_space - unreserved_space);
+            if (unreserved_space < required_maximum_available_space)
+                need_to_move.emplace(disk, required_maximum_available_space - unreserved_space);
         }
     }
 
+    auto current_time = time(nullptr);
+
     for (const auto & part : data_parts)
     {
         String reason;
         /// Don't report message to log, because logging is excessive
         if (!can_move(part, &reason))
             continue;
 
+        const auto ttl_entries_end = part->storage.move_ttl_entries_by_name.end();
+        auto best_ttl_entry_it = ttl_entries_end;
+        time_t max_max_ttl = 0;
+        for (auto & [name, ttl_info] : part->ttl_infos.moves_ttl)
+        {
+            auto move_ttl_entry_it = part->storage.move_ttl_entries_by_name.find(name);
+            if (move_ttl_entry_it != part->storage.move_ttl_entries_by_name.end())
+            {
+                if (ttl_info.max < current_time && max_max_ttl < ttl_info.max)
+                {
+                    best_ttl_entry_it = move_ttl_entry_it;
+                    max_max_ttl = ttl_info.max;
+                }
+            }
+        }
+        if (best_ttl_entry_it != ttl_entries_end)
+        {
+            auto & move_ttl_entry = best_ttl_entry_it->second;
+            if (move_ttl_entry.destination_type == ASTTTLElement::DestinationType::VOLUME)
+            {
+                auto volume_ptr = policy->getVolumeByName(move_ttl_entry.destination_name);
+                if (volume_ptr)
+                {
+                    auto reservation = volume_ptr->reserve(part->bytes_on_disk);
+                    if (reservation)
+                    {
+                        parts_to_move.emplace_back(part, std::move(reservation));
```

Comment:
What if `can_move == false`?

Reply:
Updated link: https://github.com/ClickHouse/ClickHouse/pull/7420/files/7a171b46f59b7fecd5286161c83a3a3aa199c47b#r337920407
In this case we will go into this code. And part won't be moved by any of policy or table TTL.